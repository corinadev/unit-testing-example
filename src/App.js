import React, { Component } from 'react';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import { withStyles } from '@material-ui/core/styles';

// Higher-Order Component
export const withState = (Wrapped) => {
  return class extends Component {
    state = { bool: true }
    toggle = () => this.setState({bool: !this.state.bool});
    render(){
      return <Wrapped bool={this.state.bool} toggle={this.toggle} {...this.props} />
    }
  }
}

// Dumb component
export const App = ({classes, bool, toggle}) => {
  return (
    <div className={classes.something}>
      <Typography className={classes.anotherThing}>Boolean value is: {bool ? 'true' : 'false'}</Typography>
      <Button className={classes.button} onClick={toggle}>Change boolean</Button>
    </div>
  )
}

// styles for CSS in JS
const styles = theme => ({
  something: {
    display: "flex",
    flexDirection: "column",
    [theme.breakpoints.up("md")]: {
      flexDirection: "row"
    }
  },
  button: {
    width: "100%",
    color: "blue"
  }  
});

// Creating final container with styles and state
const StyledComponent = withStyles(styles, {withTheme: true})(App);
const Container = withState(StyledComponent);
export default Container;
