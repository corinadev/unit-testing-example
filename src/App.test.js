import React from 'react';
import ReactDOM from 'react-dom';
import { App } from './App';
import './enzyme';
import { withState } from './App';
import { shallow } from 'enzyme';
import { constants } from 'perf_hooks';
import renderer from 'react-test-renderer';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';

// Arrange
const BaseButton = ({ name }) => (<button>Hello {name}!</button>);
const ButtonWithState = withState(BaseButton);

describe('wrap a component', () => {
  let theButton;

  beforeEach(() => {
    // Act
    theButton = shallow(<ButtonWithState name={"John"} />);
  });

  // Assert
  it('should contain a `bool` property in the state', () => {
    expect(theButton.state()).toEqual({ "bool": true });
  });

  it('should pass base components props forward', () => {
    expect(theButton.props().name).toEqual("John");
  });
  
  it('should render the same thing as the base component', () => {
    const baseComponent = shallow(<BaseButton name={"John"} />);
    expect(baseComponent.html()).toEqual(theButton.html());
  });

  describe('.. after toggle', () => {
    beforeEach(() => {
      // Act
      // Code review - consider renaming toggle to onToggle, so you can use simulate here, which would fit better
      // theButton.simulate('toggle');
      theButton.props().toggle();
    });


    it('should change the value of bool to false', () => {
      expect(theButton.state()).toEqual({ "bool": false });
    });
  })
});

describe('App', () => {
  it('snapshot when true', () => {
    const app = renderer.create(<App bool={true} classes={{something: "someClass" }} />).toJSON();
    expect(app).toMatchSnapshot();
  });

  it('snapshot when false', () => {
    const app = renderer.create(<App bool={false} classes={{something: "someClass" }} />).toJSON();
    expect(app).toMatchSnapshot();
  });


  let wrapper, mock;

  beforeEach(() => {
    mock = jest.fn();
    wrapper = shallow(<App bool={false} toggle={mock} classes={{something: "someClass" }}/>);
  });

  it('should display the value of `bool`', () => {
    expect(wrapper.find(Typography).childAt(1).text()).toEqual("false");
  });

  it('should call `toggle` when button is clicked', () => {
    const button = wrapper.find(Button);
    button.simulate('click');
    expect(mock).toHaveBeenCalled();
  });
});