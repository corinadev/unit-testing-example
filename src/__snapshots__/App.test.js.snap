// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot when false 1`] = `
<div
  className="someClass"
>
  <p
    className="MuiTypography-root-1 MuiTypography-body1-10"
  >
    Boolean value is: 
    false
  </p>
  <button
    className="MuiButtonBase-root-41 MuiButton-root-25"
    disabled={false}
    onBlur={[Function]}
    onClick={undefined}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-26"
    >
      Change boolean
    </span>
    <span
      className="MuiTouchRipple-root-44"
    />
  </button>
</div>
`;

exports[`App snapshot when true 1`] = `
<div
  className="someClass"
>
  <p
    className="MuiTypography-root-1 MuiTypography-body1-10"
  >
    Boolean value is: 
    true
  </p>
  <button
    className="MuiButtonBase-root-41 MuiButton-root-25"
    disabled={false}
    onBlur={[Function]}
    onClick={undefined}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-26"
    >
      Change boolean
    </span>
    <span
      className="MuiTouchRipple-root-44"
    />
  </button>
</div>
`;
